sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java

```diff
@@ -292,13 +345,29 @@ private void exportBatches(ArrayList<ReadableSpan> spanList) {
       return Collections.unmodifiableList(spanDataBuffer);
     }
 
-    // Exports the list of Span protos to all the ServiceHandlers.
-    private void onBatchExport(List<SpanData> spans) {
-      // In case of any exception thrown by the service handlers continue to run.
+    // Exports the list of SpanData to the SpanExporter.
+    private void onBatchExport(final List<SpanData> spans) {
+      Future<?> submission =
+          executorService.submit(
+              new Runnable() {
+                @Override
+                public void run() {
+                  // In case of any exception thrown by the service handlers catch and log.
+                  try {
+                    spanExporter.export(spans);
+                  } catch (Throwable t) {
+                    logger.log(Level.WARNING, "Exception thrown by the export.", t);
+                  }
+                }
+              });
       try {
-        spanExporter.export(spans);
-      } catch (Throwable t) {
-        logger.log(Level.WARNING, "Exception thrown by the export.", t);
+        // wait at most for the configured timeout.
+        submission.get(exporterTimeoutMillis, TimeUnit.MILLISECONDS);
+      } catch (InterruptedException | ExecutionException e) {
+        logger.log(Level.WARNING, "Exception thrown by the export.", e);
+      } catch (TimeoutException e) {
+        logger.log(Level.WARNING, "Export timed out. Cancelling execution.", e);
+        submission.cancel(true);
```

Comment:
IIRC, this only signals the related thread that it's supposed to stop, so we need for the interrupted state in the submitted runnable above.

Reply:
I believe this is covered by sending in the `true` value for the `mayInterruptIfRunning` parameter. From the javadoc:
```
If the task has already started, then the mayInterruptIfRunning parameter determines whether the thread executing this task should be interrupted in an attempt to stop the task.
```

Reply:
Unless you're pointing out something more subtle that I'm not aware of?

Reply:
So basically `mayInterruptIfRunning` only signals the `Thread` it has been interrupted, i.e.

1) If `false`, the task will be canceled if it hasn't been started.
2) If `true`, same as 1),  **and** `Thread.currentThread().isInterrupted()` will be set to `true` from within the task, if it's already running (which then can be consulted to finish it).

See: https://stackoverflow.com/questions/1418033/java-executors-how-can-i-stop-submitted-tasks

(Actually not so obvious as indeed the javadoc give the impression it will *actually* interrupt the task ;( ).

Reply:
But, in our case, we don't have any interrupted exceptions we can catch. That will either be inside the exporter code (geez, I sure hope not), or inside the ThreadPoolExecutor itself (which I sure hope handles it correctly!).  

So, I'm not actually sure there's anything we can do to *actually* cancel long running tasks. We can just do our best to cancel them, and hope that the underlying exporter isn't sitting blocking on I/O for a long time (which, interrupting the thread isn't going to help with anyway!)

Is there something specific you were thinking we could actually do to mitigate this?

Reply:
> (which I sure hope handles it correctly!)

I just tested it and it happens what I described above (just to be super clear - in case that's what you mean). 

> We can just do our best to cancel them

That sounds good, for now at least (we can iterate later on if/when needed). But I'd like to have this as part of the documentation, just to be on the (relatively) side point (a mention that we will **try** to cancel it, etc).

(As I said, we can later try to do something if we really want to **force** cancelation ;) )

Thanks again for looking into this!

Reply:
Yeah, I think if we wanted to make a stronger assertion about cancelling long-running exports, I think we'd need to change the exporter API, and even then, someone can block indefinitely on I/O and we'd be stuck (until project Loom!). 

Thanks for looking deeply at this. I think in practice this hopefully will never end up being a real issue, or might point at a bug in the specific exporter if it does.

Reply:
>  That will either be inside the exporter code (geez, I sure hope not),

From what I gather, this is exactly what happens though. After all, if the exporter is already running, that's the only thing that is running on this thread. If you want to stop the export, then in absence of another way of the exporter to query for cancellation, using Java Thread interruption is the only way that can possibly work.

Reply:
If your vendor's exporter code has Thread.sleep() in it, or doesn't have a way to specify IO timeouts, may I recommend another vendor?  ;)  

But, anyway, the unit test shows we handle this just fine.

Reply:
Yeah, https://github.com/open-telemetry/opentelemetry-java/pull/728/files#diff-5d25410061d3d100b226740e0605e934R244 shows that indeed an InterruptedException is can be thrown inside the exporter. This could also happen in real exporters, e.g. maybe they use `Object.wait()`. But if you believe that real exporters won't be interrupted, then what is the guidance for exporter writers? Check `isInterrupted`? Have a separate timeout (ideally lower than the one for the processor)?
What I wonder is, if this feature was worth the added complexity, since a separate I/O timeout seems to be required anyway.

Reply:
I didn't say that won't be interrupted. I just meant that it was unlikely, assuming normal "map and send" sorts of operations that should live in exporters, that there was likely to be much chance that people would explicitly be surrounding their code with InterruptedException checks. That will be being handled by the executor itself that is executing the runnable that we submit.

That being said, the New Relic reporter spawns off its own thread anyway, and returns immediately. That's what I would recommend to people, in general. So, yes this does feel like it might be more complexity than it's worth; perhaps we should get the spec changed to just push this responsibility onto the exporter.